src/experimental/stick_sequence.rs

```diff
@@ -0,0 +1,174 @@
+use rand::SeedableRng;
+use rand_xoshiro::Xoshiro256Plus;
+#[cfg(feature = "serde1")]
+use serde::{Deserialize, Serialize};
+use std::sync::{Arc, RwLock};
+
+// use super::sticks_stat::StickBreakingSuffStat;
+use crate::prelude::UnitPowerLaw;
+use crate::traits::*;
+
+// We'd like to be able to serialize and deserialize StickSequence, but serde can't handle
+// `Arc` or `RwLock`. So we use `StickSequenceFmt` as an intermediate type.
+#[cfg(feature = "serde1")]
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde1", serde(rename_all = "snake_case"))]
+struct StickSequenceFmt {
+    breaker: UnitPowerLaw,
+    inner: _Inner,
+}
+
+#[cfg(feature = "serde1")]
+impl From<StickSequenceFmt> for StickSequence {
+    fn from(fmt: StickSequenceFmt) -> Self {
+        Self {
+            breaker: fmt.breaker,
+            inner: Arc::new(RwLock::new(fmt.inner)),
+        }
+    }
+}
+
+#[cfg(feature = "serde1")]
+impl From<StickSequence> for StickSequenceFmt {
+    fn from(sticks: StickSequence) -> Self {
+        Self {
+            breaker: sticks.breaker,
+            inner: sticks.inner.read().map(|inner| inner.clone()).unwrap(),
+        }
+    }
+}
+
+// NOTE: We currently derive PartialEq, but this (we think) compares the
+// internal state of the RNGs, which is probably not what we want.
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde1", serde(rename_all = "snake_case"))]
+#[derive(Clone, Debug, PartialEq)]
+pub struct _Inner {
+    rng: Xoshiro256Plus,
+    ccdf: Vec<f64>,
+}
+
+impl _Inner {
+    fn new(seed: Option<u64>) -> _Inner {
+        _Inner {
+            rng: seed.map_or_else(
+                Xoshiro256Plus::from_entropy,
+                Xoshiro256Plus::seed_from_u64,
+            ),
+            ccdf: vec![1.0],
+        }
+    }
+
+    fn extend<B: Rv<f64> + Clone>(&mut self, breaker: &B) -> f64 {
+        let p: f64 = breaker.draw(&mut self.rng);
+        let remaining_mass = self.ccdf.last().unwrap();
+        let new_remaining_mass = remaining_mass * p;
+        self.ccdf.push(new_remaining_mass);
+        new_remaining_mass
+    }
+
+    fn extend_until<B, F>(&mut self, breaker: &B, p: F)
+    where
+        B: Rv<f64> + Clone,
+        F: Fn(&_Inner) -> bool,
+    {
+        while !p(self) {
+            self.extend(breaker);
+        }
+    }
+}
+
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde1", serde(rename_all = "snake_case"))]
+#[derive(Clone, Debug)]
+pub struct StickSequence {
```

Comment:
You'll need to use `#[serde(from = "FromType")]` and `#[serde(into = "IntoType")]` to get this to work.

```rust
#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde1", serde(rename_all = "snake_case", from="StickSequenceFmt", into="StickSequenceFmt"))]
#[derive(Clone, Debug)]
pub struct StickSequence {
//...
}
```

This is required to work in Lace. Please prioritize this over tests.